joint_trajectory_controller/test/joint_trajectory_controller_test.cpp

```diff
@@ -289,14 +361,82 @@ class JointTrajectoryControllerTest : public ::testing::Test
     if(!start_controller.response.ok) return false;
     return true;
   }
+
+  static ros::Duration getTrajectoryDuration(const trajectory_msgs::JointTrajectory &traj)
+  {
+    return traj.points.back().time_from_start;
+  }
+
+  static bool checkActionGoalState(const ActionClientPtr& action_client,
+                                   const actionlib::SimpleClientGoalState& state)
+  {
+    return action_client->getState() == state;
+  }
+
+  static bool waitForActionGoalState(const ActionClientPtr& action_client,
+                                     const actionlib::SimpleClientGoalState& state,
+                                     const ros::Duration& timeout = ros::Duration(TIMEOUT_ACTION_RESULT_S))
+  {
+    return waitForEvent(std::bind(checkActionGoalState, action_client, state),
+                        "action goal state " + state.getText(),
+                        timeout);
+  }
+
+  static bool checkActionResultErrorCode(const ActionClientPtr& action_client,
+                                         const control_msgs::FollowJointTrajectoryResult::_error_code_type& error_code)
+  {
+    return action_client->getResult()->error_code == error_code;
+  }
+
+  static AssertionResult waitForEvent(const std::function<bool()>& check_event,
+                                      const std::string& event_description,
+                                      const ros::Duration& timeout,
+                                      unsigned int repeat = 1)
+  {
+    unsigned int count{0};
+    ros::Time start_time{ros::Time::now()};
+    while (ros::ok())
+    {
+      count = check_event() ? (count+1) : 0;
+      if ((ros::Time::now() - start_time) > timeout)
+      {
+        return AssertionFailure() << "Timed out after " << timeout.toSec() << "s waiting for "
+                                  << event_description << ".";
+      }
+      ros::Duration(0.1).sleep();
+      if (count == repeat)
+      {
+        return AssertionSuccess();
+      }
+    }
+    return AssertionFailure() << "ROS shutdown.";
+  }
+
+  static bool trajectoryPointsAlmostEqual(const trajectory_msgs::JointTrajectoryPoint& p1,
```

Comment:
How about making them global functions? Maybe some of these static functions could also be used in other tests. So it would even make sense to put them into a separate header?

Reply:
love the idea!

Reply:
Thanks! I will push another commit.